### DISCUSSION TITLE
Custom code rendering

### DISCUSSION THREAD

**Discussion by matkoch** [facebook/docusaurus#6447] [Q&A] [answered]
Hi everyone!

I'm maintaining an OSS .NET project and it took me forever to research a better platform to deliver documentation. Now that I've found Docusaurus, I feel almost embarrassed seeing how popular it is, but again, it's probably because I have absolutely no connection to frontend technology.

Docusaurus seems to have absolutely everything I wished for and more. The only thing I have no clue how to accomplish is custom code rendering. So instead of using regular code fences and specifying the language, I'd like to provide my own `<pre><code>` fragments that already have `<span class="keyword">` and other custom classes, along with my custom CSS.

I've found another discussion #3009 and realized I could do `<span class={ ... }>`, which seemed like a good start, but apparently, I lose features like the copy button and file titles.

Any help would be greatly appreciated.

**Comment by Josh-Cena** [comment, marked as answer]
Hi! Docusaurus uses prism-react-renderer to render your code. If you want another renderer, you can swizzle `CodeBlock` and re-implement it whichever way you want. https://deploy-preview-6296--docusaurus-2.netlify.app/docs/using-themes/

**Reply by matkoch** [reply]
Wow, that was fast 😅 Thanks for the pointer! Could you give me another hint on how I can render `code` as HTML and not let it be escaped?

**Reply by Josh-Cena** [reply]
`<code>` is usually already preserved as HTML. In what case is it "escaped"?

**Reply by matkoch** [reply]
With my first try I did the following in markdown file:
```
``​`jsx title="src/pages/my-react-page.js"
<span>Foo</span>
`​``
```

In the `CodeBlock` component I removed all the parts involving the token mapping, and ended with (line 81):

```
<code className={styles.codeBlockLines}>{code}</code>
```

This however will not treat `code` as HTML content. So I still see:

<img width="313" alt="Firefox-Create a Page  My Site-2022-01-23-xTzeq7dv" src="https://user-images.githubusercontent.com/5005566/150664059-4d68840b-2e8f-4a46-a3d3-4f09adaf7982.png">

**Reply by Josh-Cena** [reply]
Yes, because code blocks from Markdown would always pass its content as plain string. I suggest you pass the string you actually want to render in Markdown:

````md
<!-- whatever you actually intend to render, without the markup -->
``​`jsx title="src/pages/my-react-page.js"
Foo
`​``
````

And then do your magic in the `CodeBlock`:

```jsx
// Takes a code string and returns the parsed tokens with styles as JSX
function render(code) {
  return <span className="foo">{code}</span>;
}

const renderedCode = render(code);

<code className={styles.codeBlockLines}>{renderedCode}</code>
```

I hope that makes sense?

**Reply by matkoch** [reply]
In this case unfortunately not. The classified code (with spans) is generated from another executable, and I don't think I want to embed calling this into a react component (I guess it wouldn't even work). So while this is definitely not beautiful, I really need the HTML fragments inside the code fence, and then let it render unescaped.

As an example, here is the HTML that the command-line tool generates: https://jsfiddle.net/eut2kphm/121/

_Thanks btw for being so responsive. I did not expect that :)_

**Reply by Josh-Cena** [reply]
Okay, if you actually have the code pre-rendered and available as plain strings, you can use the JSX counterpart because HTML tags in Markdown code blocks cannot be escaped (for good reason):

https://docusaurus.io/docs/next/markdown-features/code-blocks#using-jsx-markup

The `CodeBlock` component will let you pass a title. However, in order to implement copy, you need to swizzle `CodeBlock` and tell it how to extract plain strings from the bunch of markup.

**Reply by matkoch** [reply]
Shouldn't this work then in an mdx file?

```
import CodeBlock from '@theme/CodeBlock';

<CodeBlock>
<span class="foo">Foo</span>
</CodeBlock>
```

**Reply by Josh-Cena** [reply]
Yes, it does. Nit: should be `className="foo"`

**Reply by matkoch** [reply]
I'm getting an error though (crashed page) :(

> content.replace is not a function

Even with the other example from the docs:
```
<pre>
  <b>Input: </b>1 2 3 4{'\n'}
  <b>Output: </b>"366300745"{'\n'}
</pre>
```

I end up with: 
<img width="276" alt="Firefox-markdown-features  My Site-2022-01-23-6DRiONxU" src="https://user-images.githubusercontent.com/5005566/150665192-f8713e7e-16a7-4c52-b7eb-b2ff37cef097.png">

**Reply by Josh-Cena** [reply]
If you use JSX in your code block, you have to get rid of a lot of processing we have done because they are done on the assumption that the content is plain string. This has been improved in the next version. Since the infrastructure has changed quite a bit in the next version, I suggest you [use canary version](https://docusaurus.io/community/canary) and re-swizzle your component.

**Reply by Josh-Cena** [reply]
See https://github.com/facebook/docusaurus/pull/6177
